### PR TITLE
Change refine imports behaviour for qualified imports

### PR DESCRIPTION
**Pull request by joyfulmantis**
This PR addresses issue #3770, specifically, it implements [option 2](https://github.com/haskell/haskell-language-server/issues/3770#issuecomment-1716058934) "If an import is qualified, refine it and keep the same qualification, but don't give it an explicit import list."

### CODE REVIEW

File: plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs [hunk 1]
```diff
@@ -476,16 +476,20 @@ filterByImport _ _ = Nothing
 
 constructImport :: LImportDecl GhcRn -> (ModuleName, [AvailInfo]) -> LImportDecl GhcRn
 #if MIN_VERSION_ghc(9,5,0)
-constructImport (L lim imd@ImportDecl {ideclName = L _ _, ideclImportList = Just (hiding, L _ names)})
+constructImport (L lim imd@ImportDecl {ideclName = L _ _, ideclQualified = qualified, ideclImportList = Just (hiding, L _ names)})
```
Comment (michaelpj) [inline plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs:479]: I'm confused - it looks like this case only triggers in the case that the original import *had* an import list! Otherwise `ideclImportList` would be `Nothing` and so we would just go into the fallthrough.

The original bug report suggested that we were converting imports without import lists to imports with import lists, but I can't see how that would happen from this logic. Does the test definitely fail without these changes?

Comment (michaelpj) [reply plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs:479]: Because as it stands, it looks to me like it will take something like this:
```
import RefineA as A (x, y)
```
and turn it into
```
import RefineB as A (x)
import RefineC as A (y)
```
which actually seems fine.

Comment (michaelpj) [reply plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs:479]: (It would also be nice to have a tests for the case of qualified-but-with-import-list also, even if it's weird)

**Review by michaelpj — COMMENTED** [review]
I'm confused about whether the bug even happens!

File: plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs (hunk 1 above)
Comment (joyfulmantis) [reply plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs:479]: It's because on line 343 we actually calculate the refined imports from the generated minimal imports list, not the actual imports list. This is all legacy stuff, but I could see why they would want to do that if they assumed people would always want an explicit import list for refined imports (It certainly helps to know why we are importing some module)

Comment (joyfulmantis) [reply plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs:479]: > Because as it stands, it looks to me like it will take something like this:
> 
> ```
> import RefineA as A (x, y)
> ```
> 
> and turn it into
> 
> ```
> import RefineB as A (x)
> import RefineC as A (y)
> ```
> 
> which actually seems fine.

That also means that in this case we will end up dumping the explicit import list.

Comment (michaelpj) [reply plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs:479]: From what you're saying, it sounds like the minimal imports code action will also put import lists on qualified imports? That _also_ seems questionable to me - in that situation you can see where something is coming from by looking at the qualifier of the usage, e.g.

```
import Foo qualified as F

F.whatever
```
turning it into 
```
import Foo qualified as F (whatever)

F.whatever
```
Seems worse IMO.

I guess an alternative question is: what if we processed the non-minimal imports in refine imports? Then it seems like we would handle the second case I suggested, since it would be looking at the original imports? It seems reasonable to me for these two things to be orthogonal. WDYT?

Comment (joyfulmantis) [reply plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs]: For minimal imports, we specifically filter out qualified imports (line 324 in this PR). I moved some of the logic around, so now refine imports can work on the original import. For some reason, the edge case is still causing problems (There is a test case for it now, and you should be able to see it failing). Not sure why though.

File: plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs [hunk 2]
```diff
@@ -494,5 +497,7 @@ constructImport (L lim imd@ImportDecl{ideclName = L _ _, ideclHiding = Just (hid
           containsAvail name avail =
             any (\an -> printOutputable an == (printOutputable . ieName . unLoc $ name))
               $ availNamesWithSelectors avail
-
-constructImport lim _ = lim
+          noExplicitList mList = case mList of
+                                      Nothing      -> True
+                                      Just (_, xs) -> not $ null xs
```
Comment (michaelpj) [inline plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs:502]: This is a super weird case. This is handling `import Foo qualified ()`, right? I guess we might as well handle it, but I think it would be also reasonable to treat that as an explicitly specified import list, even if it's empty.

File: plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs [hunk 3]
```diff
@@ -474,18 +473,22 @@ filterByImport (L _ ImportDecl{ideclHiding = Just (_, L _ names)})
...
 #else
-constructImport (L lim imd@ImportDecl{ideclName = L _ _, ideclHiding = Just (hiding, L _ names)})
+constructImport ImportDecl{ideclQualified = qualified, ideclHiding = origHiding} imd@ImportDecl{ideclHiding = Just (hiding, L _ names)}
 #endif
-  (newModuleName, avails) = L lim imd
+  (newModuleName, avails) = imd
     { ideclName = noLocA newModuleName
 #if MIN_VERSION_ghc(9,5,0)
-    , ideclImportList = Just (hiding, noLocA newNames)
+    , ideclImportList = if noExplicitList origHiding && qualified /= NotQualified
```
Comment (michaelpj) [inline plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs:485]: TBH, at this point why not just put an import list in exactly if the original import had one? Then we avoid changing the status as much as possible. i.e. this condition could just be `isNothing origHiding`?

**Review by michaelpj — APPROVED** [review]
I like the direction of  travel, just one more potential simplification.

File: plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs (hunk 3 above)
Comment (joyfulmantis) [reply plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs]: Yeah, I was overthinking things. We only need isNothing and that fixes the test case too.
